File: bindings/php/tests/BoxTest.php
<?php

declare(strict_types=1);

use PHPUnit\Framework\TestCase;

/**
 * crypto_box - no official vector exists for this composite construction (same posture as
 * crypto_secretbox, D-51/D-169's own "Provenance" section) - correctness (round trip, including a
 * message larger than the 25-byte KEM payload), rejection (tampered wire segments, wrong key),
 * misuse (wrong-length/invalid key encodings, truncated input).
 */
final class BoxTest extends TestCase
{
    public function testRoundTripsSealOpen(): void
    {
        $secretKey = dstu_core_box_keygen();
        $publicKey = dstu_core_box_public_key($secretKey);
        $message = "a message for the public key's holder only";
        $sealed = dstu_core_box_seal($publicKey, $message);
        $this->assertSame($message, dstu_core_box_open($secretKey, $sealed));
    }

    public function testHandlesAnEmptyMessage(): void
    {
        $secretKey = dstu_core_box_keygen();
        $publicKey = dstu_core_box_public_key($secretKey);
        $sealed = dstu_core_box_seal($publicKey, '');
        $this->assertSame('', dstu_core_box_open($secretKey, $sealed));
    }

    public function testHandlesMessageLargerThanTheKemPayload(): void
    {
        $secretKey = dstu_core_box_keygen();
        $publicKey = dstu_core_box_public_key($secretKey);
        $message = str_repeat('x', 10_000);
        $sealed = dstu_core_box_seal($publicKey, $message);
        $this->assertSame($message, dstu_core_box_open($secretKey, $sealed));
    }

    public function testTwoSealsUseDifferentEphemeralMaterial(): void
    {
        $secretKey = dstu_core_box_keygen();
        $publicKey = dstu_core_box_public_key($secretKey);
        $message = 'same message twice';
        $this->assertNotSame(
            dstu_core_box_seal($publicKey, $message),
            dstu_core_box_seal($publicKey, $message)
        );
    }

    public function testRejectsTamperedKemPrefix(): void
    {
        $secretKey = dstu_core_box_keygen();
        $publicKey = dstu_core_box_public_key($secretKey);
        $sealed = dstu_core_box_seal($publicKey, 'message');
        $sealed[0] = chr(ord($sealed[0]) ^ 1);
        $this->expectException(DstuCoreException::class);
        dstu_core_box_open($secretKey, $sealed);
    }

    public function testRejectsTamperedCiphertext(): void
    {
        $secretKey = dstu_core_box_keygen();
        $publicKey = dstu_core_box_public_key($secretKey);
        $sealed = dstu_core_box_seal($publicKey, 'message');
        $sealed[strlen($sealed) - 1] = chr(ord($sealed[-1]) ^ 1);
        $this->expectException(DstuCoreException::class);
        dstu_core_box_open($secretKey, $sealed);
    }

    public function testRejectsWrongSecretKey(): void
    {
        $secretKey = dstu_core_box_keygen();
        $publicKey = dstu_core_box_public_key($secretKey);
        $otherSecretKey = dstu_core_box_keygen();
        $sealed = dstu_core_box_seal($publicKey, 'message');
        $this->expectException(DstuCoreException::class);
        dstu_core_box_open($otherSecretKey, $sealed);
    }

    public function testRejectsWrongLengthSecretKey(): void
    {
        $this->expectException(\ValueError::class);
        dstu_core_box_public_key('too short');
    }

    public function testRejectsZeroSecretKey(): void
    {
        $this->expectException(\ValueError::class);
        dstu_core_box_public_key(str_repeat("\x00", 32));
    }

    public function testRejectsWrongLengthPublicKey(): void
    {
        $this->expectException(\ValueError::class);
        dstu_core_box_seal('too short', 'message');
    }

    public function testRejectsDegeneratePublicKeyX(): void
    {
        $this->expectException(\ValueError::class);
        dstu_core_box_seal(str_repeat("\x00", 32), 'message'); // x = 0
    }

    public function testRejectsTruncatedSealedInput(): void
    {
        $secretKey = dstu_core_box_keygen();
        $this->expectException(DstuCoreException::class);
        dstu_core_box_open($secretKey, 'short');
    }
}

File: bindings/go/dstu/box_test.go
package dstu

import (
	"bytes"
	"testing"
)

// BoxSecretKey/BoxPublicKey (crypto_box) - no official vector exists for this composite
// construction (same posture as SecretboxKey, D-51/D-169's own "Provenance" section) -
// correctness (round trip, including a message larger than the 25-byte KEM payload), rejection
// (tampered wire segments, wrong key), misuse (wrong-length/invalid key encodings, truncated
// input).

func TestBoxSealOpenRoundTrips(t *testing.T) {
	secretKey, _ := GenerateBoxSecretKey()
	defer secretKey.Close()
	publicKey := secretKey.PublicKey()
	defer publicKey.Close()
	message := []byte("a message for the public key's holder only")
	sealed, err := publicKey.Seal(message)
	if err != nil {
		t.Fatal(err)
	}
	got, err := secretKey.Open(sealed)
	if err != nil {
		t.Fatal(err)
	}
	if !bytes.Equal(got, message) {
		t.Fatalf("got %q, want %q", got, message)
	}
}

func TestBoxSealHandlesEmptyMessage(t *testing.T) {
	secretKey, _ := GenerateBoxSecretKey()
	defer secretKey.Close()
	publicKey := secretKey.PublicKey()
	defer publicKey.Close()
	sealed, err := publicKey.Seal(nil)
	if err != nil {
		t.Fatal(err)
	}
	got, err := secretKey.Open(sealed)
	if err != nil {
		t.Fatal(err)
	}
	if len(got) != 0 {
		t.Fatalf("got %d bytes, want 0", len(got))
	}
}

func TestBoxSealHandlesMessageLargerThanTheKemPayload(t *testing.T) {
	secretKey, _ := GenerateBoxSecretKey()
	defer secretKey.Close()
	publicKey := secretKey.PublicKey()
	defer publicKey.Close()
	message := bytes.Repeat([]byte("x"), 10_000)
	sealed, err := publicKey.Seal(message)
	if err != nil {
		t.Fatal(err)
	}
	got, err := secretKey.Open(sealed)
	if err != nil {
		t.Fatal(err)
	}
	if !bytes.Equal(got, message) {
		t.Fatal("round trip mismatch for a message larger than the KEM payload")
	}
}

func TestBoxTwoSealsUseDifferentEphemeralMaterial(t *testing.T) {
	secretKey, _ := GenerateBoxSecretKey()
	defer secretKey.Close()
	publicKey := secretKey.PublicKey()
	defer publicKey.Close()
	message := []byte("same message twice")
	a, _ := publicKey.Seal(message)
	b, _ := publicKey.Seal(message)
	if bytes.Equal(a, b) {
		t.Fatal("two seals produced identical output")
	}
}

func TestBoxTamperedKemPrefixIsRejected(t *testing.T) {
	secretKey, _ := GenerateBoxSecretKey()
	defer secretKey.Close()
	publicKey := secretKey.PublicKey()
	defer publicKey.Close()
	sealed, _ := publicKey.Seal([]byte("message"))
	sealed[0] ^= 1
	if _, err := secretKey.Open(sealed); err == nil {
		t.Fatal("expected an error")
	}
}

func TestBoxTamperedCiphertextIsRejected(t *testing.T) {
	secretKey, _ := GenerateBoxSecretKey()
	defer secretKey.Close()
	publicKey := secretKey.PublicKey()
	defer publicKey.Close()
	sealed, _ := publicKey.Seal([]byte("message"))
	sealed[len(sealed)-1] ^= 1
	if _, err := secretKey.Open(sealed); err == nil {
		t.Fatal("expected an error")
	}
}

func TestBoxWrongSecretKeyIsRejected(t *testing.T) {
	secretKey, _ := GenerateBoxSecretKey()
	defer secretKey.Close()
	publicKey := secretKey.PublicKey()
	defer publicKey.Close()
	otherSecretKey, _ := GenerateBoxSecretKey()
	defer otherSecretKey.Close()
	sealed, _ := publicKey.Seal([]byte("message"))
	if _, err := otherSecretKey.Open(sealed); err == nil {
		t.Fatal("expected an error")
	}
}

func TestBoxWrongLengthSecretKeyIsRejected(t *testing.T) {
	if _, err := BoxSecretKeyFromBytes(make([]byte, 10)); err == nil {
		t.Fatal("expected an error")
	}
}

func TestBoxZeroSecretKeyIsRejected(t *testing.T) {
	if _, err := BoxSecretKeyFromBytes(make([]byte, BoxSecretKeyBytes)); err == nil {
		t.Fatal("expected an error")
	}
}

func TestBoxWrongLengthPublicKeyIsRejected(t *testing.T) {
	if _, err := BoxPublicKeyFromBytes(make([]byte, 10)); err == nil {
		t.Fatal("expected an error")
	}
}

func TestBoxDegeneratePublicKeyXIsRejected(t *testing.T) {
	if _, err := BoxPublicKeyFromBytes(make([]byte, BoxPublicKeyBytes)); err == nil { // x = 0
		t.Fatal("expected an error")
	}
}

func TestBoxTruncatedSealedInputIsRejected(t *testing.T) {
	secretKey, _ := GenerateBoxSecretKey()
	defer secretKey.Close()
	if _, err := secretKey.Open([]byte("short")); err == nil {
		t.Fatal("expected an error")
	}
}
